### ISSUE TITLE
Implementing lora-ctl with webui-forge

### ISSUE THREAD

**Issue by Omegastick** [lllyasviel/stable-diffusion-webui-forge#68]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits

### What would your feature do ?

I'd like to get the [sd-webui-loractl](https://github.com/cheald/sd-webui-loractl) extension working with webui-forge. It works with the latest Automatic1111 commit, so I'm guessing one of the more central changes here is making it incompatible.

I'm open to forking (or rewriting) it myself, but I'd appreciate any guidance on where to start.

### Proposed workflow

1. Install extension
2. Activate in WebUI
3. Use <lora:lora_name:0.0@0,1.0@1> syntax

**Comment by mohitsrivastavazzz** [comment]
by any chance do you use civitai helper extention and is that working for you in this fork?

**Comment by sashasubbbb** [comment]
I agree, this is a much needed extenstion.

**Comment by Omegastick** [comment]
> by any chance do you use civitai helper extention and is that working for you in this fork?

Sorry, I don't use that extension.

**Comment by rafstahelin** [comment]
second that feature request

**Comment by brentjohnston** [comment]
I would also love to see this extension working in this, It's so underrated. It's really awesome being able to control at what step a too strong lora kicks in and then kicks out, or mix loras however you want. Here was a recent post that covered it https://old.reddit.com/r/StableDiffusion/comments/1aqlvi0/psa_dont_ignore_the_sdwebuiloractl_extension_for/

**Comment by cheald** [comment]
Hi there. I'm the author of loractl. I've been out of the SD scene for a bit and don't have bandwidth to adapt loractl to Forge, but it's fundamentally a monkeypatch on A1111's network handling code. It's doing two specific things:

1. First, it [registers itself as the handler for the `lora` extra network when it's enabled](https://github.com/cheald/sd-webui-loractl/blob/master/scripts/loractl.py#L28-L36). This enables it to take over parsing for those <lora:...> blocks. A1111 automatically feeds them to the extension.
2. Second, it [patches the network.Network class](https://github.com/cheald/sd-webui-loractl/blob/master/loractl/lib/network_patch.py#L45) by hijacking its `te_multiplier` and `unet_multiplier` properties, and replacing them with a properties which actually calls a function in loractl to compute the te/unet multipliers (the lora weight) instead. Those functions consider the current generation step as a part of the value they return.

A1111, on each step, checks to see which extra networks are in play, and at what weights. Fudging the te_multiplier/unet_multiplier weights has the effect of causing A1111 to unapply the previous lora weights and reapply new ones at the new weights given.

All the heavy lifting is done by A1111 - all loractl is really doing is changing the te_multiplier and unet_multiplier values on a per-step basis, which just happens to make A1111 do the right thing.

From a quick gander at the Forge network handling, [te/unet weights are calculated during network activation](https://github.com/lllyasviel/stable-diffusion-webui-forge/blob/d81e353d8928147bbd973068d0efbb2802affe0f/extensions-builtin/Lora/extra_networks_lora.py#L12-L41), but rather than being taken from overridable properties, they're taken directly from the `params` parameter passed to `activate`. To make the loractl concept work, `ExtraNetworkLora#activate` would have to be monkeypatched with a replacement implentation which would parse the params list with the multi-step parser (that could be lifted directly from [loractl's utils](https://github.com/cheald/sd-webui-loractl/blob/master/loractl/lib/utils.py#L39)) then compute the te/unet weights based on the keyframes parsed from the params, the current step, and the total number of desired steps, like [loractl does here](https://github.com/cheald/sd-webui-loractl/blob/master/loractl/lib/network_patch.py#L14-L15).

In theory it shouldn't be too difficult, but it would probably need to be a fork of loractl, since it would be targeting a fundamentally different codebase.

**Comment by lllyasviel** [comment]
thanks a lot. we will take a look soon about lora system

**Comment by kyle215ps3** [comment]
@lllyasviel Thank you!!! This is absolutely needed!!! It's amazing being able to force the initial generation to place the character first before the LoRa does its work.
It's even better when paired with BREAK keywords!

**Comment by davizca** [comment]
Thanks a lot! Will be very handy to have dynamic weighting in LORA's in Forge!

**Comment by kyle215ps3** [comment]
I third this! :D

**Comment by BurnZeZ** [comment]
> A1111, on each step, checks to see which extra networks are in play, and at what weights. Fudging the te_multiplier/unet_multiplier weights has the effect of causing A1111 to unapply the previous lora weights and reapply new ones at the new weights given.
@cheald Is this computationally expensive?

**Comment by cheald** [comment]
Yes, it's why loractl tanks it/s during periods that lora weights are changing step by step.

**Comment by Ppepepe2** [comment]
Any news on the matter? this exstention would so awesome

**Comment by altoiddealer** [comment]
> thanks a lot. we will take a look soon about lora system

@lllyasviel  Hello - I see you are very busy, with very important things!

You flagged this issue as "High Priority" - I just wanted to remind you :)

I don't know if this has any hope to fit into your plans, but it would be so nice.

Keep up the amazing work!  It is very exciting to have you back

**Comment by moudahaddad14** [comment]
**any new news on this extension**, it might be a game changer for loras in forge tbh! 🤩

**Comment by altoiddealer** [comment]
Trying to make progress towards this [on my branch here](https://github.com/altoiddealer/stable-diffusion-webui-forge-altoids/tree/integrate_loractl) with the intention that loractl would be implemented as a **builtin-extension**

From what little I've done so far, it seems to get hooked in with the Lora networks and dynamically calculate the weights based on the prompt syntax, exactly as it does in A1111.

It's just a matter of those taking effect....

@Panchovix pinging you because you seemed to have a better understanding of Forge LORA inner workings based on your comments here

https://github.com/Panchovix/stable-diffusion-webui-reForge/issues/36

**Comment by altoiddealer** [comment]
Guys - I did not so much as even attempt to make any further progress with this.

### BUT

@Panchovix DID make the dream happen in ReForge.  He has yet to say whether he has ambitions to push it to Forge or not, or comment on whether it is even possible.

I'm just writing this so everyone in love with Forge memory handling, and the original loractl extension - can NOW enjoy that with ReForge.

https://github.com/Panchovix/stable-diffusion-webui-reForge

https://github.com/Panchovix/stable-diffusion-webui-reForge/issues/36

**Comment by Athari** [comment]
Nothing seems to be happening. 😟 Super-advanced ultra-fast lora-ctl implementation in reForge is nice, but I'd rather have something slow in Forge.

**Comment by EleVicted** [comment]
Patched. Tested, works great with multiple LoRAs!

@https://github.com/cheald/sd-webui-loractl/issues/38#issuecomment-3868678427
